executor/joiner.go

```diff
@@ -215,6 +261,32 @@ func (j *semiJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chu
 	return false, false, nil
 }
 
+func (j *semiJoiner) tryToMatchOuters(outers chunk.Iterator, inner chunk.Row, chk *chunk.Chunk, outerRowStatus []outerRowStatusFlag) (_ []outerRowStatusFlag, err error) {
+	outerRowStatus = outerRowStatus[:0]
+	outer, numToAppend := outers.Current(), chk.RequiredRows()-chk.NumRows()
+	if len(j.conditions) == 0 {
+		for ; outer != outers.End() && numToAppend > 0; outer, numToAppend = outers.Next(), numToAppend-1 {
+			chk.AppendPartialRow(0, outer)
+			outerRowStatus = append(outerRowStatus, outerRowMatched)
+		}
+		return outerRowStatus, nil
+	}
+	for outer := outers.Current(); outer != outers.End() && numToAppend > 0; outer, numToAppend = outers.Next(), numToAppend-1 {
+		j.makeShallowJoinRow(j.outerIsRight, inner, outer)
+		// For SemiJoin, we can safely treat null result of join conditions as false,
+		// so we ignore the nullness returned by EvalBool here.
+		matched, _, err := expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
+		if err != nil {
+			return outerRowStatus, err
+		}
+		outerRowStatus = append(outerRowStatus, outerRowMatched)
+		if matched {
+			chk.AppendPartialRow(0, outer)
```

Comment:
Should we only decrease `numToAppend` when this operation is performed?

Reply:
No~ outers.Next is needed.